src/CodeGen_D3D12Compute_Dev.cpp

```diff
@@ -896,22 +896,79 @@ void CodeGen_D3D12Compute_Dev::CodeGen_D3D12Compute_C::add_kernel(Stmt s,
         constants[i].size += constants[i - 1].size;
     }
 
-    // Find all the shared allocations and declare them at global scope.
-    class FindSharedAllocations : public IRVisitor {
-        using IRVisitor::visit;
-        void visit(const Allocate *op) override {
-            op->body.accept(this);
+    // Find all the shared allocations, uniquify their names,
+    // and declare them at global scope.
+    class FindSharedAllocationsAndUniquify : public IRMutator {
+        using IRMutator::visit;
+        Stmt visit(const Allocate *op) override {
             if (is_shared_allocation(op)) {
-                allocs.push_back(op);
+                // Because these will go in global scope,
+                // we need to ensure they have unique names.
+                std::string new_name = unique_name(op->name);
+                replacements[op->name] = new_name;
+
+                std::vector<Expr> new_extents;
+                for (size_t i = 0; i < op->extents.size(); i++) {
+                    new_extents.push_back(mutate(op->extents[i]));
```

Comment:
Do we need to mutate these things? Can't we just pass op->extents directly to Allocate::make?

Reply:
It seems extremely unlikely (impossible?) that the size of one shared allocation would depend on a load from another shared allocation. Still, safer to just always mutate than to make assumptions that may change in the future, especially if the cost of mutation is low (which it should be, as extent is probably a constant int).

Reply:
Probably, but I just wanted to be safer & reflect the logic in `IRMutator::visit(const Allocate*)`.